feat: Medbay cryo pods

Carbonhell opened the pull request:
<!-- The text between the arrows are comments - they will not be visible on your PR. -->
<!-- Please read these guidelines before opening your PR: https://docs.spacestation14.io/en/getting-started/pr-guideline -->
Fixes #11245 

## About the PR <!-- Describe the Pull Request here. What does it change? What other things could this impact? -->
This PR adds a new machine called cryo pods (or cryo cells). This is a medical machine which allows cooling down a patient to extremely low temperatures, without damaging them. This is useful paired with medicines that react on low temperatures such as cryoxadone. The machine also accepts a solution container to draw chemicals from, periodically injecting them in the mob at a specific rate (1 unit per second right now).
The machine requires a small atmos setup to work, a power amount slightly lower than clone pods and it can be tampered in several ways.

1) The machine requires an atmos setup to prepare an air mix at an acceptable pressure for mobs (one atmosphere) and with a low temperature (which can be achieved with the use of a freezer). Forgetting the pressure requirement can make the medical machine become a torture machine instead, quickly killing the patient with overpressurization (or due to no pressure aswell). The machine overrides the cold protection of the patient to 10°K.

2) Normally you can insert a mob (or yourself) by dragging it into the pod. Ejection works through the use of a verb. Ejection might not work if the pod is locked. This can be achieved by cutting a specific wire (mending solves it) or emagging it (which requires the machine to be rebuilt).

3) The machine circuitboard can be obtained by high level research (same as clone pods). This PR does not affects map (and I think those pods should be built rather than mapped in).

4) Once the patient is ejected, the cold protection is nullified. This means damage will start to stack. This is intended, medics can heal the tiny damage caused by this or prevent it somehow (heating the room, giving clothes, dunno).

The PR introduces two (somewhat?) important code tweaks. One is the ChangeDamageGetThresholdsEvent ref event which allows to override the temperature threshold similar to how it's done for the AtmosExposedGetAirEvent event. Another is a force argument flag to the Stand method of the StandingStateSystem to ignore overrides (this is needed to allow showing the mob contained within the cryo pod in certain cases).

**Screenshots**
<!-- If applicable, add screenshots to showcase your PR. If your PR is a visual change, add
screenshots or it's liable to be closed by maintainers. -->
Example of an atmos setup (the cryo pod itself can also be scanned with a gas analyzer to ensure everything's ok)
![image](https://user-images.githubusercontent.com/11891037/190642329-356c3d0d-1621-40da-b28e-76b0884c5e29.png)

No power, the human inside is visible, floating within the pod.
![image](https://user-images.githubusercontent.com/11891037/190642503-5ea23195-7d4a-49e3-b14b-f775f7f1bcb1.png)

Power available, the pod turns on automatically and begins to inject the human inside with chemicals
![image](https://user-images.githubusercontent.com/11891037/190642590-796dccf6-f3f8-4b4c-8551-e858077b0954.png)

Interacting with the pod with an empty hand shows a shamelessly recycled health analyzer UI to monitor the patient with. The cryo pod in the picture also has its maintenance pod open.
![image](https://user-images.githubusercontent.com/11891037/190643100-a00b6bb8-772f-41f5-ad4b-9ba43a23390a.png)

The locking mechanism was tampered with and the ejection verb doesn't work.
![image](https://user-images.githubusercontent.com/11891037/190643328-2b5e8d01-6526-49d6-bc38-b56488a5bdde.png)



**Changelog**
<!--
Here you can fill out a changelog that will automatically be added to the game when your PR is merged
There are 4 icons for changelog entries: add, remove, tweak, fix. I trust you can figure out the rest.

You can put your name after the :cl: symbol to change the name that shows in the changelog (otherwise it takes your GitHub username)
Like so: :cl: PJB

Generally, only put things in changelogs that players actually care about. Stuff like "Refactored X system, no changes should be visible" shouldn't be on a changelog.

For writing actual entries, don't consider the entry type suffix (e.g. add) to be "part" of the sentence:
bad: - add: a new tool for engineers
good: - add: added a new tool for engineers
-->

:cl:
- add: Medbay can now obtain cryo pods, specialized cryogenics machines used to lower the temperature of a patient to extreme values somewhat safely to use specific chemicals, such as cryoxadone.

lapatison commented:
Just FYI, clonexadone has been removed from the game in #11253

Carbonhell commented:
Ah I missed that PR, thanks for telling me!

On a sidenote, there's a slight graphical glitch that happens when the pod's open and someone walks on the tile below it. I left it in because otherwise people who get ejected would appear under the pod, which IMO is uglier. Not sure if there's a way to solve this.
![image](https://user-images.githubusercontent.com/11891037/190650221-7dca8888-de38-40d0-b20c-7db592328756.png)

Also, I just rebased on the latest commit and removed the deprecated GasAnalyzable component.

Kromabae commented:
> Just FYI, clonexadone has been removed from the game in #11253

it was removed, but would probably be added again if it has a use, and it was removed becuase it's exactly the same as cryoxadone
if the chems just get different effects and such they can both exist again, no?

EmoGarbage404 commented:
> Ah I missed that PR, thanks for telling me!
> 
> On a sidenote, there's a slight graphical glitch that happens when the pod's open and someone walks on the tile below it. I left it in because otherwise people who get ejected would appear under the pod, which IMO is uglier. Not sure if there's a way to solve this. ![image](https://user-images.githubusercontent.com/11891037/190650221-7dca8888-de38-40d0-b20c-7db592328756.png)
> 
> Also, I just rebased on the latest commit and removed the deprecated GasAnalyzable component.

Probably would want to give the top the proper drawdepth and then just have the game deal with drawdepth on the pod portion when exiting. Doesnt climbing on the pod put it on top of it visually?

ShadowCommander commented:
> On a sidenote, there's a slight graphical glitch that happens when the pod's open and someone walks on the tile below it. I left it in because otherwise people who get ejected would appear under the pod, which IMO is uglier. Not sure if there's a way to solve this.

Unless it's been changed, the other medical machines have half-wall collision and are climbable. When the person is ejected they are automatically put into the climbing state.

This makes it collidable while not pushing players out.

theashtronaut commented:
Your gas analyzer scan looks like it uses the old gas analyzer before I reworked it. Looks like the cryo pod has a node container, but if you want to actually specify what mixes the gas analyzer displays, like having it instead display the gas mixture in the Cryo pod Component, you need to add a handler for GasAnalyzerScanEvent in the system. You can look at portable scrubbers for an example on this.

metalgearsloth commented on Content.Server/Medical/CryoPodSystem.cs:
```diff
@@ -0,0 +1,359 @@
+    public override void Update(float frameTime)
+    {
+        base.Update(frameTime);
+        foreach (var cryoPod in EntityQuery<CryoPodComponent>())
+        {
+            if (!cryoPod.Enabled)
+            {
+                continue;
+            }
```
This should just use an active component.

metalgearsloth commented on Content.Server/Medical/CryoPodSystem.cs:
```diff
@@ -0,0 +1,359 @@
+            var container = _itemSlotsSystem.GetItem(cryoPod.Owner, "beakerSlot");
+            if (container != null
+                && container.Value.Valid
+                && cryoPod.BodyContainer.ContainedEntity != null
+                && TryComp<BloodstreamComponent>(cryoPod.BodyContainer.ContainedEntity, out var bloodstream)
```
```suggestion
                && TryComp<BloodstreamComponent>(cryoPod.BodyContainer.ContainedEntity, out var bloodstream)
```

Should also be using entityqueries.

Carbonhell commented on Content.Server/Temperature/Systems/TemperatureSystem.cs:
```diff
@@ -172,29 +172,35 @@ private void ChangeDamage(EntityUid uid, TemperatureComponent temperature)
+            var heatDamageThreshold = ev.HeatDamageThreshold ?? temperature.HeatDamageThreshold;
+            var coldDamageThreshold = ev.ColdDamageThreshold ?? temperature.ColdDamageThreshold;
+
+            if (temperature.CurrentTemperature >= heatDamageThreshold)
```
Instead of doing it specifically for the cryo pods, I added a new ContainerTemperatureDamageThresholdsComponent to allow a parent container to override the temperature thresholds of a contained entity by simply adding the component in YAML. Not sure if it's overengineering but I think there might be more use cases for this and it isn't really expensive or anything, I think.

Carbonhell commented on Content.Server/Medical/CryoPodSystem.cs:
```diff
@@ -0,0 +1,367 @@
+    {
+        if (args.Powered)
+        {
+            EnsureComp<ActiveCryoPodComponent>(uid);
```
Is this the proper way to add an Active component to an entity? In my initial tests, I used AddComp instead of EnsureComp. I tried to delete an entity from the debug verb and I had a weird runtime, "Attempted to add a ActiveCryoPodComponent component to an entity (cryopod) while it is terminating" (from AddComponentInternal). I didn't manage to replicate it, but I didn't see any special check being done elsewhere to prevent this case, or maybe I failed to find one.

vulppine commented on Content.Server/Medical/CryoPodWireAction.cs:
```diff
@@ -0,0 +1,59 @@
+        StatusLightState lightState = StatusLightState.Off;
+        if (EntityManager.TryGetComponent<CryoPodComponent>(wire.Owner, out var cryoPodComponent) && cryoPodComponent.Locked)
+        {
+            lightState = StatusLightState.On; //TODO figure out why this doesn't get updated when the pod is emagged
```
might need to add a check in wires for emag so that it refreshes the state of the status lights (out of scope for this PR)

vulppine commented on Content.Server/Medical/CryoPodWireAction.cs:
```diff
@@ -0,0 +1,59 @@
+/// Causes a failure in the cryo pod ejection system when cut. A crowbar will be needed to pry open the pod.
+/// </summary>
+[DataDefinition]
+public sealed class CryoPodWireAction: BaseWireAction
```
could change the name of this to be something like CryoPodEjectLockWireAction, this is a little vague

vulppine commented on Content.Server/Medical/CryoPodSystem.cs:
```diff
@@ -0,0 +1,367 @@
+            return;
+        }
+
+        var doAfterArgs = new DoAfterEventArgs(args.User, cryoPodComponent.EntryDelay, default, args.Dragged)
```
replace the target parameter here (the default you placed down) with the EntityUid of the cryopod, and replace the broadcast finished event with a targeted event instead

vulppine commented on Content.Server/Medical/CryoPodSystem.cs:
```diff
@@ -0,0 +1,367 @@
+                return;
+            cryoPodComponent.IsPrying = true;
+
+            _toolSystem.UseTool(args.Used, args.User, uid, 0f,
```
could convert this to use targeted events, instead? the form of these events lends itself towards that - see the UseTool function for more information

to summarize, you just have to define the doAfter event target, so that once the doafter finishes, it knows to send the event as a targeted event to the doAfter event target - and in this system, you'd just change it so that the event listener is instead listening to a targeted event, and you wouldn't have to add the uid of the cryopod and the reference to the cryopod component in the event

vulppine commented on Content.Server/Medical/CryoPodSystem.cs:
```diff
@@ -0,0 +1,367 @@
+        if (args.Handled || !cryoPodComponent.Locked || cryoPodComponent.BodyContainer.ContainedEntity == null)
+            return;
+
+        if (EntityManager.TryGetComponent(args.Used, out ToolComponent? tool)
```
```suggestion
        if (TryComp(args.Used, out ToolComponent? tool)
```

vulppine commented on Content.Server/Medical/CryoPodSystem.cs:
```diff
@@ -0,0 +1,367 @@
+        {
+            return;
+        }
+        if (!EntityManager.TryGetComponent(args.User, out ActorComponent? actor))
```
```suggestion
        if (!TryComp(args.User, out ActorComponent? actor))
```

Carbonhell commented:
Thank you @theashtronaut , I've implemented the event handler as you stated, I decided to show both the cryo pod air and the port air since scrubbers do something similar and it makes sense for cryo pods aswell.
Thank you very much @vulppine for the review, I've addressed all the comments and I made all the events targeted since it makes no sense for any to be a broadcast.

vulppine reviewed (approved):
looks fine, might want another maintainer to review too

mirrorcult commented on Content.Server/Medical/CryoPodSystem.cs:
```diff
@@ -0,0 +1,389 @@
+        SubscribeLocalEvent<CryoPodComponent, CryoPodPryInterrupted>(OnCryoPodPryInterrupted);
+        SubscribeLocalEvent<CryoPodComponent, DestructionEventArgs>(OnDestroyed);
+        SubscribeLocalEvent<CryoPodComponent, GasAnalyzerScanEvent>(OnGasAnalyzed);
+
```
```suggestion
```

mirrorcult commented on Content.Server/Medical/CryoPodSystem.cs:
```diff
@@ -0,0 +1,389 @@
+    [Dependency] private readonly ToolSystem _toolSystem = default!;
+    [Dependency] private readonly SharedAppearanceSystem _appearanceSystem = default!;
+    [Dependency] private readonly MobStateSystem _mobStateSystem = default!;
+
```
```suggestion
```

mirrorcult commented on Content.Server/Medical/CryoPodSystem.cs:
```diff
@@ -0,0 +1,389 @@
+                && patient != null
+                && _solutionContainerSystem.TryGetFitsInDispenser(container.Value, out var containerSolution))
+            {
+
+                var bloodstreamQuery = GetEntityQuery<BloodstreamComponent>();
```
this doesnt actually change anything since you're not hoisting it out of a loop, just use a regular `TryComp`

mirrorcult commented on Content.Server/Medical/CryoPodSystem.cs:
```diff
@@ -0,0 +1,389 @@
+        InsertBody(uid, args.ToInsert, cryoPodComponent);
+    }
+
+    private void HandleInteractHand(EntityUid uid, CryoPodComponent cryoPodComponent, InteractHandEvent args)
+    {
+        if (!HasComp<ActiveCryoPodComponent>(uid))
+        {
+            return;
+        }
+        if (!TryComp(args.User, out ActorComponent? actor))
+            return;
+
+        if (cryoPodComponent.BodyContainer.ContainedEntity != null)
+        {
+            if (args.User == cryoPodComponent.BodyContainer.ContainedEntity)
+            {
+                return;
+            }
+            _userInterfaceSystem.TryOpen(uid, SharedHealthAnalyzerComponent.HealthAnalyzerUiKey.Key, actor.PlayerSession);
+            _userInterfaceSystem.TrySendUiMessage(
+                uid,
+                SharedHealthAnalyzerComponent.HealthAnalyzerUiKey.Key,
+                new SharedHealthAnalyzerComponent.HealthAnalyzerScannedUserMessage(cryoPodComponent.BodyContainer.ContainedEntity));
+            args.Handled = true;
+        }
+    }
```
use the `ActivatableUI` component in yaml for this instead to avoid duplication. you can subscribe to some 'after activatable UI open' (i think) event to do the UI message sending, and theres a before UI open attempt event to cancel if its not active etc

Carbonhell commented on Content.Server/Medical/Components/CryoPodComponent.cs:
```diff
@@ -0,0 +1,71 @@
+    /// <summary>
+    /// How many units to transfer per tick from the beaker to the mob?
+    /// </summary>
+    public float BeakerTransferAmount = 1f;
```
I've made everything a datafield except IsPrying which I think doesn't make sense to keep between a save+load event.

mirrorcult replied:
sure

mirrorcult commented:
code looks good. on testing ingame some notes:
- 1 canister is too high pressure for the cryopod at 73.15 with a small pipenet. i was still able to heal just fine, but something to note i suppose. nothing necessarily has to be done about this
- people are fully conscious while within the cryopod. IMO they shouldnt be able to eject themselves from it, and it might be reasonable to just put them in a sleeping state?

theashtronaut commented:
> Regarding 2, I was fairly conflicted on this point. I have a lot of bad experiences from ss13 where I'd be forgotten alone in a pod due to it putting the patient to sleep. I would really prefer to avoid causing the same experience to other players. If someone wants to use the cryo pod maliciously, the locking mechanism enabled either by an emag or by pulsing the right wire works. If it ends up being too easy for people to patch themself with a cryopod, though, it can easily be patched in a future PR.

The fix that was put in place for the "medical forgot about you" was that the cryopod will auto-eject patient that are finished and beep angerly when doing it to alert medical (not continuously, just when ejected). I think tying the auto-eject function to the hacked/wire side could be good, so a malicious player can disable the auto-eject feature should they want. It would let people perma-trap others though. Maybe there is also an auto-ejection after X minutes like 5 or 10 minutes, where it dumps the patient out eventually even if hacked just to stop perma-lock from happening? That way you would have plenty of time for nefarious deeds, but can't just shove people into a pod somewhere in maintenance and remove them from the game at the same time?

Carbonhell commented:
> The fix that was put in place for the "medical forgot about you" was that the cryopod will auto-eject patient that are finished and beep angerly when doing it to alert medical (not continuously, just when ejected). I think tying the auto-eject function to the hacked/wire side could be good, so a malicious player can disable the auto-eject feature should they want. It would let people perma-trap others though. Maybe there is also an auto-ejection after X minutes like 5 or 10 minutes, where it dumps the patient out eventually even if hacked just to stop perma-lock from happening? That way you would have plenty of time for nefarious deeds, but can't just shove people into a pod somewhere in maintenance and remove them from the game at the same time?

I don't really see much difference between auto-ejection when fully healed vs letting the medic or the patient eject when they want. Also perma-trapping is already possible for antagonistic reason by either cutting the lock wire or by the use of an emag, but since patients do not fall asleep you can at least have the chance to scream for help, which is a fair balance imo. If you build a cryo pod in maintenance to lock people up, with the long do-after there is in place you would definitely need the victim cuffed, at that point you could remove them from the game in any other way as well, so I do not see a balance issue in this way.
I do not see any obvious balance pain point in the current implementation, my idea still is to get this PR in as is and see how the community uses the pods and balance them out if some unexpected OP use shows up.

Carbonhell commented:
I resolved the conflicts along with some small errors regarding refactors to ItemSlotsSystem (GetItem is now GetItemOrNull) and Body stuff (In the CanInsert method, I just had to swap SharedBodyComponent to BodyComponent). I've tested ingame and everything works.

I've noticed something bizarre though: the CO2 buildup generated by a patient breathing within a cryo pod is enormous: 0.08 mols of CO2 were being generated every gas analyzer UI refresh. This means that, with an atmos setup without a working gas filter, the patient will die of poison extremely quickly.
I've tried an isolated test in a 1x1 room and I noticed my mob (a lizard) was breathing out 0.02 mols of CO2. I'm not sure why it's x4 in a cryo pod, but 0.02 seems to also be too much according to mirrorcult.

Carbonhell commented:
There's only one issue left regarding cryo pods with a wall above them: you can enter the pod, you cannot eject yourself from the pod though since the wall somehow stops you. Also, if someone ejects you, you'll be pushed above the wall.
![image](https://user-images.githubusercontent.com/11891037/198276642-0e86724b-783e-4ec8-8612-cbe11f0ceec3.png)

I'm not really sure how to handle this case, I'll need to think about it a bit more.

Carbonhell commented:
![cryo_pod_working_2](https://user-images.githubusercontent.com/11891037/198733261-511bffac-b28d-4dca-81b5-9457314ea5b5.png)
I refactored the code regarding the sprite offsetting when inside the cryopod to be handled by editing the clientside SpriteComponent Offset field, instead of editing the transform serverside, therefore the issues related to my last comments are now fixed. The PR should be good to go if all the tests pass.

rolfero commented on Content.Client/Medical/Cryogenics/CryoPodSystem.cs:
```diff
@@ -0,0 +1,71 @@
+    {
+        base.Initialize();
+
+        SubscribeLocalEvent<InsideCryoPodComponent, ComponentAdd>(OnCryoPodInsertion);
```
I think the ComponentAdd event runs when the components are added to the entity, which means it doesn't guarantee that the SpriteComponent has been added yet.

I think MapInitEvent is better instead, which guarantees that all components have been added AND initialized.

metalgearsloth replied:
It's pre-serialization data and unless you really need it (outside of engine you likely don't) you're better off using startup.

Carbonhell replied:
I've changed it to ComponentStartup, thank you!

rolfero commented on Content.Server/Medical/CryoPodSystem.cs:
```diff
@@ -0,0 +1,384 @@
+    private void OnComponentInit(EntityUid uid, CryoPodComponent cryoPodComponent, ComponentInit args)
+    {
+        base.Initialize();
+        cryoPodComponent.BodyContainer = _containerSystem.EnsureContainer<ContainerSlot>(uid, $"scanner-bodyContainer");
```
"scanner-bodyContainer" does not need to be an interpolated string, right?
```suggestion
        cryoPodComponent.BodyContainer = _containerSystem.EnsureContainer<ContainerSlot>(uid, "scanner-bodyContainer");
```

rolfero commented on Content.Server/Medical/Components/CryoPodComponent.cs:
```diff
@@ -0,0 +1,85 @@
+    [DataField("solutionContainerName")]
+    public string SolutionContainerName { get; set; } = "beakerSlot";
+
+    /// <summary>
+    /// Accumulator used to keep track of when to inject chemicals from the inserted beaker to the mob inside the pod.
+    /// </summary>
+    [DataField("accumulator")]
+    public float Accumulator = 0f;
```
I think this instead should be turned into a target TimeSpan to avoid accumulators. See my comment in the CryoPodSystem update() method.

rolfero commented on Content.Server/Medical/CryoPodSystem.cs:
```diff
@@ -0,0 +1,384 @@
+        base.Update(frameTime);
+        foreach (var (_, cryoPod) in EntityQuery<ActiveCryoPodComponent, CryoPodComponent>())
+        {
+            cryoPod.Accumulator += frameTime;
+
+            if (cryoPod.Accumulator < cryoPod.BeakerTransferTime)
+                continue;
+
+            cryoPod.Accumulator -= cryoPod.BeakerTransferTime;
```
This should be turned into a target timespan instead, to avoid the use of accumulators. You would need to the IGameTiming dependency, and then use it like this:

```suggestion
            if (_gameTiming.CurTime < cryoPod.TargetTime)
                continue;

            cryoPod.TargetTime = cryoPod.TargetTime + TimeSpan.FromSeconds(cryoPod.BeakerTransferTime);
```

rolfero replied:
Annnnd this should prob also take into account time paused, so
if (_gameTiming.CurTime < cryoPod.TargetTime **+ time entity paused** ) continue;

Carbonhell replied:
What exactly do you mean by time entity paused? I saw that _gameTiming has a Paused bool but I haven't seen it used in a way similar to what you said anywhere, do you have an example? Thanks in advance

metalgearsloth commented on Content.Shared/Medical/Cryogenics/InsideCryoPodComponent.cs:
```diff
@@ -0,0 +1,13 @@
+
+namespace Content.Shared.Medical.Cryogenics;
+
+[RegisterComponent]
+[NetworkedComponent]
+public sealed class InsideCryoPodComponent: Component
+{
+    [ViewVariables]
+    public EntityUid Holder;
+
+    public Vector2 PreviousOffset { get; set; } = new(0, 0);
+}
```
Doesn't look like this handles serialization, same with the container code in cryopodsystem, so if someone saves a map with something inside a cryo pod it will break.

brainfood1183 replied:
this still being worked on? shame if we dont eventually see this finished.

Carbonhell replied:
Yeah I am, I have an exam soon and I need to sort that out first. After that, I need to figure out how to move the interactions in shared as suggested (some of them like drag&drop use doafters and I remember that causes problems since they aren't predicted yet?). I also need to figure out what exactly needs to be done to handle serialization (simply datafield it? I need to understand how serialization works at all here before continuing). And the only issue left after this is the ParentColdDamageThreshold override, i'm really not sure how to handle recursive parent changes. I need to see if there's already something else that does this (most likely yes).
I expect to be able to work on this again on this weekend at worst, sorry for the wait!

Carbonhell replied:
I've added datafields for those two properties, but I don't understand how exactly I should handle containers for the cryo pod. As far as I saw, other instances have a NonSerializable decorator on ContainerSlot fields. Other than that I didn't see any custom logic that handles serialization/deserialization, could you give me an example?

metalgearsloth reviewed (changes requested):
Would be nicer if the interactions were in shared.

Carbonhell commented on Content.Server/Medical/CryoPodSystem.cs:
```diff
@@ -0,0 +1,384 @@
+        };
+        _doAfterSystem.DoAfter(doAfterArgs);
+        args.Handled = true;
+    }
```
Yep, it was. Fixed! Thanks.

Carbonhell commented on Content.Server/Medical/CryoPodSystem.cs:
```diff
@@ -0,0 +1,384 @@
+
+    private void UpdateAppearance(EntityUid uid, CryoPodComponent? cryoPod = null)
+    {
+        if (!Resolve(uid, ref cryoPod))
+            return;
+
+        var cryoPodEnabled = HasComp<ActiveCryoPodComponent>(uid);
+        _appearanceSystem.SetData(uid, SharedCryoPodComponent.CryoPodVisuals.ContainsEntity, cryoPod.BodyContainer.ContainedEntity == null);
+        _appearanceSystem.SetData(uid, SharedCryoPodComponent.CryoPodVisuals.IsOn, cryoPodEnabled);
+        if (TryComp<PointLightComponent>(uid, out var light))
+        {
+            light.Enabled = cryoPodEnabled && cryoPod.BodyContainer.ContainedEntity != null;
+        }
+
+        _appearanceSystem.SetData(uid,SharedCryoPodComponent.CryoPodVisuals.PanelOpen, false);
```
I used the system because SetData on the component was deprecated: https://github.com/space-wizards/RobustToolbox/blob/master/Robust.Shared/GameObjects/Components/Appearance/AppearanceComponent.cs#L28
Isn't this the correct way?

Macoron replied:
I guess, @metalgearsloth meant that you can resolve `AppearanceComponent` in the beginning of the function and use it as parameter for `SetData`. This is small optimization that saves three resolves inside `SetData` body.

Carbonhell commented on Content.Server/Medical/InsideCryoPodSystem.cs:
```diff
@@ -0,0 +1,57 @@
+        {
+            base.Initialize();
+            // Atmos overrides
+            SubscribeLocalEvent<InsideCryoPodComponent, InhaleLocationEvent>(OnInhaleLocation);
+            SubscribeLocalEvent<InsideCryoPodComponent, ExhaleLocationEvent>(OnExhaleLocation);
+            SubscribeLocalEvent<InsideCryoPodComponent, AtmosExposedGetAirEvent>(OnGetAir);
+
+            SubscribeLocalEvent<InsideCryoPodComponent, DownAttemptEvent>(HandleDown);
```
Right now it's handled in CryoPodSystem, where this component is ensured on the patient entity when inserting him: https://github.com/space-wizards/space-station-14/pull/11349/files#diff-1bea9cdee7424b2ae962d2615216fc1313c504e2a64195a77b973d0c79949057R145

And it is removed when the patient is ejected: https://github.com/space-wizards/space-station-14/pull/11349/files#diff-1bea9cdee7424b2ae962d2615216fc1313c504e2a64195a77b973d0c79949057R177
The latter also happens if the cryo pod is destroyed or deconstructed (through the DestructionEventArgs listener), which I believe should cover all the cases. I assume a failsafe on InsideCryoPodComponent which removes itself from the entity if the parent container isn't a cryo pod anymore would be safer, I'll address that.

metalgearsloth replied:
Yeah so if any other piece of code ejects it rn it would leak so you should also handle container removals. Also I think we have a destructible trigger to dump container contents which is probably better rather than subscribing to destruction event.

Carbonhell replied:
Done, through the EntGotRemovedFromContainerMessage event listener. Should be the correct way to handle this if I understood container events correctly. I also found the trigger you were talking about, although it seems to have no containers whitelist so now the machine components also get ejected, but i think fixing that is outside the scope of the PR since it happens with other machines too.

Carbonhell commented on Content.Server/Medical/Components/CryoPodComponent.cs:
```diff
@@ -0,0 +1,85 @@
+    [DataField("pryDelay")]
+    public float PryDelay = 5f;
+
+    /// <summary>
+    /// If true, the eject verb will not work on the pod and the user must use a crowbar to pry the pod open.
+    /// </summary>
+    [ViewVariables(VVAccess.ReadWrite)]
+    [DataField("locked")]
+    public bool Locked { get; set; }
+
+    /// <summary>
+    /// Causes the pod to be locked without being fixable by messing with wires.
+    /// </summary>
+    [ViewVariables(VVAccess.ReadWrite)]
+    [DataField("permaLocked")]
+    public bool PermaLocked { get; set; }
```
I wasn't aware of LockComponent, thank you. I just checked the code, but this component seems to be related to user-facing locks which can depend on access restriction and can be toggled at will by an user, and they also have an already implemented action on emagged which simply removes the lock component. The cryo pod Locked var is a lot simpler and acts as a simple flag to prevent ejection methods from working. There seems to be nothing I can reuse from that component, and due to the different behaviour (user-facing lock offered by LockComponent vs a "manual" lock that strictly depends on the cryo pod implementation) I think it makes more sense to have it implemented this way, I'd like a second opinion though.

metalgearsloth commented on Content.Server/Medical/Components/CryoPodComponent.cs:
```diff
@@ -0,0 +1,63 @@
+namespace Content.Server.Medical.Components;
+
+[RegisterComponent]
+[ComponentReference(typeof(SharedCryoPodComponent))]
```
Shouldn't use compreferences unless absolutely necessary (they're not in this instance).

Carbonhell replied:
Are you sure they're not necessary? When removing them, I get this error: "Component is not a valid reference type: SharedCryoPodComponent". I think it makes sense since the shared component is abstract yet the shared system code registers listeners on it, so the reference to the shared component has to exist somehow, no?

metalgearsloth replied:
You just need to update subscribers to use client / server versions but using the shared in the method signature is still fine.

metalgearsloth commented on Content.Server/Medical/Components/CryoPodComponent.cs:
```diff
@@ -0,0 +1,63 @@
+    [DataField("beakerTransferTime")]
+    public float BeakerTransferTime = 1f;
+
+    [ViewVariables(VVAccess.ReadWrite)]
+    [DataField("nextInjectionTime")]
+    public TimeSpan? NextInjectionTime;
```
Should this use TimeOffsetSerializer?

metalgearsloth commented on Content.Server/Medical/CryoPodSystem.cs:
```diff
@@ -0,0 +1,236 @@
+        base.Update(frameTime);
+        foreach (var (_, cryoPod) in EntityQuery<ActiveCryoPodComponent, CryoPodComponent>())
+        {
+            if (_gameTiming.CurTime < cryoPod.NextInjectionTime)
```
CurTime should be stored in a local variable as it has to be re-calculated every time.

metalgearsloth commented on Content.Server/Medical/CryoPodSystem.cs:
```diff
@@ -0,0 +1,236 @@
+                continue;
+            cryoPod.NextInjectionTime = _gameTiming.CurTime + TimeSpan.FromSeconds(cryoPod.BeakerTransferTime);
+
+            var container = _itemSlotsSystem.GetItemOrNull(cryoPod.Owner, cryoPod.SolutionContainerName);
```
Should be using an entityquery for this + bloodstream + solutioncontainermanager.

metalgearsloth commented on Content.Server/Temperature/Systems/TemperatureSystem.cs:
```diff
@@ -209,6 +215,76 @@ private void OnTemperatureChangeAttempt(EntityUid uid, TemperatureProtectionComp
+        /// Recalculate parent thresholds for the root entity and all its descendant.
+        /// </summary>
+        /// <param name="root"></param>
+        private void RecursiveThresholdUpdate(EntityUid root)
+        {
+            RecalculateParentThresholds(root, null);
+
+            foreach (var child in Transform(root).ChildEntities)
+            {
+                RecalculateParentThresholds(child, null);
+            }
+        }
```
Should be using EntityQuerys for transform and temperature for this.

metalgearsloth commented on Content.Server/Temperature/Systems/TemperatureSystem.cs:
```diff
@@ -209,6 +215,76 @@ private void OnTemperatureChangeAttempt(EntityUid uid, TemperatureProtectionComp
             args.Args.TemperatureDelta *= component.Coefficient;
         }
+
+        private void OnParentChange(EntityUid uid, TemperatureComponent component,
+            ref EntParentChangedMessage args)
+        {
+            RecursiveThresholdUpdate(uid);
+        }
```
Isn't it possible to short-circuit this? If a mob changes parents which is frequent this is gonna go through like 40 children unnecessarily.

metalgearsloth commented on Content.Shared/Medical/Cryogenics/InsideCryoPodComponent.cs:
```diff
@@ -0,0 +1,16 @@
+[NetworkedComponent]
+public sealed class InsideCryoPodComponent: Component
+{
+    [ViewVariables]
+    [DataField("holder")]
+    public EntityUid Holder;
```
Wouldn't this just be the parent?

metalgearsloth commented on Content.Shared/Medical/Cryogenics/SharedCryoPodComponent.cs:
```diff
@@ -0,0 +1,58 @@
+    /// <summary>
+    /// Container for mobs inserted in the pod.
+    /// </summary>
+    [ViewVariables, NonSerialized]
```
Typo?

metalgearsloth commented on Content.Shared/Medical/Cryogenics/SharedCryoPodSystem.cs:
```diff
@@ -0,0 +1,185 @@
+        var cryoPodEnabled = HasComp<ActiveCryoPodComponent>(uid);
+        _appearanceSystem.SetData(uid, SharedCryoPodComponent.CryoPodVisuals.ContainsEntity, cryoPod.BodyContainer.ContainedEntity == null, appearance);
+        _appearanceSystem.SetData(uid, SharedCryoPodComponent.CryoPodVisuals.IsOn, cryoPodEnabled, appearance);
+        if (TryComp<SharedPointLightComponent>(uid, out var light))
+        {
+            light.Enabled = cryoPodEnabled && cryoPod.BodyContainer.ContainedEntity != null;
+        }
```
Appearance component isn't required for this.

metalgearsloth commented on Content.Shared/Medical/Cryogenics/SharedCryoPodSystem.cs:
```diff
@@ -0,0 +1,185 @@
+            return;
+        }
+
+        cryoPodComponent.PermaLocked = true;
```
I'm still highly suss on emag permanently locking someone in considering there's nothing they can personally do about it, I think it should just be a regular lock.

theashtronaut replied:
Given the patient is no longer put to sleep, what if people could break cryotubes (they are just fancy glass tubes after all) which would let someone trapped inside break out. It could be pretty strong so just using fists might take many hits.

Carbonhell replied:
My point of view is that for this to happen, you need to consume an emag's charge, so the effect needs to be meaningful. If the patient can just as easily get himself outside the pod, I doubt anyone would waste a charge for this. But from the patient's point of view, they still can scream and speak on the radio, which is useful considering those pods would most of the time be located in the middle of medbay. 
Since you're not asleep, you can actually get yourself outside the pod if you have a crowbar. Crowbar interaction works inside the pod too. I didn't plan on allowing this initially but I guess it can be left in to give a better survival chance to the patient.
And, as @theashtronaut said, you can already break the pod to eject someone stuck in it, but currently if you're inside the pod you cannot attack the pod itself.

metalgearsloth reviewed (changes requested):
Does this look okay with rotated eye?

Carbonhell commented:
> Does this look okay with rotated eye?

It completely broke when rotating the eye haha, thank you for making me notice. I've used only a single directional sprite now which looks okay and works when rotating the eye as well.

EmoGarbage404 reviewed (approved):
Lgtm

Carbonhell commented:
The failing test seems to be related to the wrong license used in some rsi files completely unrelated to my PR, so I believe it's safe to say those can be ignored.
